ebu_tt_live/node/deduplicator.py

```diff
@@ -0,0 +1,199 @@
+        if self.is_document(document):
+
+            if document.sequence_identifier == self._sequence_identifier:
+                raise UnexpectedSequenceIdentifierError()
+
+            if self.check_if_document_seen(document=document):
+
+                self.limit_sequence_to_one(document)
+
+                document.sequence_identifier = self._sequence_identifier
+
+                self.remove_duplication(document=document)
+                print document.get_xml()
+
+                document.validate()
+                self.producer_carriage.emit_data(data=document, **kwargs)
+
+    def remove_duplication(self, document):
+        _old_id_dict = dict({})
+        _new_id_dict = dict({})
+        hash_dict = dict({})
+
+        if document.binding.head.styling is not None and document.binding.head.layout is not None:
+            styles = document.binding.head.styling.style
+            document.binding.head.styling.style = None
+
+            regions = document.binding.head.layout.region
+            document.binding.head.layout.region = None
+
+            self.CollateUniqueVals(styles, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(styles, document.binding.head.styling.style, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            self.CollateUniqueVals(regions, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(regions, document.binding.head.layout.region, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            replace_id_refs = ReplaceStylesAndRegions(document.binding, \
+                                                      _old_id_dict, \
+                                                      _new_id_dict)
+            replace_id_refs.proceed()
+        else:
+            return None
+
+    def CollateUniqueVals(self, element_list, _old_id_dict, _new_id_dict, \
+                          hash_dict):
+        for value in element_list:
+            #deduplicating in-line styles
+            if value.style is not None:
+                for old_id_index in range(len(value.style)):
+                    old_id_ref = _old_id_dict.get(value.style[old_id_index])
+                    new_id_ref = _new_id_dict.get(old_id_ref)
+
+                    value.style[old_id_index] = new_id_ref
+
+            #deduplicating region elements
+            if value is not None:
+                unique_val = ComparableElement(value)
+                # stores references of original <xml:id> to <my_hash>
+                _old_id_dict[value.id] = unique_val.my_hash
+                # stores references of <my_hash> to <tt:region>
+                hash_dict[unique_val.my_hash] = value
+
+    def AppendNewElements(self, element_list, element_to_append_to, _old_id_dict, \
+                          _new_id_dict, hash_dict):
+        for hash_val in hash_dict:
+            new_id = hash_dict.get(hash_val)
```

Comment:
Might be more elegant to change this loop a bit along the lines of:

```python
for hash_val, new_id in hash_dict.iteritems():
```

which does the lookup for you without making another call.